### PR TITLE
Add new LSP extension for workspace symbol lookup

### LINKED ISSUE #4881: Workspace symbol search returns no results when query includes dependencies
If I search for `*HashSet`, VS Code (1.46, also observed in 1.45 before updating), it says "No matching workspace symbols", but I see non-empty results from the request

<details>
<summary>request trace</summary>

```
[Trace - 3:27:48 PM] Sending request 'workspace/symbol - (41)'.
Params: {
    "query": "*HashSet"
}


[Trace - 3:27:48 PM] Received response 'workspace/symbol - (41)' in 42ms.
Result: [
    {
        "kind": 26,
        "location": {
            "range": {
                "end": {
                    "character": 52,
                    "line": 135
                },
                "start": {
                    "character": 0,
                    "line": 133
                }
            },
            "uri": "file:///home/jbalint/.cargo/registry/src/github.com-1ecc6299db9ec823/fnv-1.0.7/lib.rs"
        },
        "name": "FnvHashSet"
    },
    {
        "kind": 26,
        "location": {
            "range": {
                "end": {
                    "character": 65,
                    "line": 46
                },
                "start": {
                    "character": 0,
                    "line": 44
                }
            },
            "uri": "file:///home/jbalint/.cargo/registry/src/github.com-1ecc6299db9ec823/rustc-hash-1.1.0/src/lib.rs"
        },
        "name": "FxHashSet"
    },
    {
        "kind": 23,
        "location": {
            "range": {
                "end": {
                    "character": 1,
                    "line": 111
                },
                "start": {
                    "character": 0,
                    "line": 16
                }
            },
            "uri": "file:///home/jbalint/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/collections/hash/set.rs"
        },
        "name": "HashSet"
    }
]
```
</details>

### PR DESCRIPTION
**Pull request by alcroito**
As well as all symbol types (functions, modules).

Remove outdated documentation regarding symbol lookup filtering.

Closes #4881

### CODE REVIEW

**Comment by alcroito** [comment]
The current PR makes the UX of using RA with VSCode a bit nicer (IMHO), but I realize it removes the filtering functionality from other editors.

If that's unacceptable, perhaps we could expose these filtering knobs as settings to be controlled in VSCode? Or detect VSCode as the LSP client and set them accordingly?

**Comment by matklad** [comment]
Yeah, it's sad that this got broken in vs code. That being said, I think searching in the current workspace by default works better, and I definitely think that we should keep the heuristic of "search for types by default, if there are no results, search for all symbols instead". 

Let's make this user configurable. In config.rs, I think we should add:

```rust
search.dependencies: bool = false,
search.allSymbols: bool = false,
```

**Comment by petr-tik** [comment]
> Let's make this user configurable. In config.rs, I think we should add:
> 
> ```rust
> search.dependencies: bool = false,
> search.allSymbols: bool = false,
> ```

Would it be possible to override these settings on per-request basis? i.e. starting ra with default to not search dependencies, but changing it for 1 request, when you might want to see them. 

I think this is the perfect usecase for an elisp function that takes an optional prefix argument similar to [this function in lsp-clangd](https://github.com/emacs-lsp/lsp-mode/blob/master/clients/lsp-clangd.el#L245-L249). That way, we can call `M-x lsp-find-workspace-symbol` or `C-u M-x lsp-find-workspace-symbol` to include dependencies in the results.

**Comment by matklad** [comment]
@petr-tik emacs can directly add `*` and `#` to the query string, this PR is specifically for VS Code.

That being said, given that `#*` workaround is broken in vscode, it make sense to design a proper LSP extension here. See the how-to for how this should be done: https://github.com/rust-analyzer/rust-analyzer/tree/master/docs/dev#how-to-

basically, we need to add two fields to lsp request params: `search_scope` and `search_kinds`

**Comment by alcroito** [comment]
Pushed code to implement a new LSP extension extending the usual workspace symbol lookup request.

Aside from bikeshedding names (suggestions welcomes), I'd like some feedback for a few open points:

1. I added the additional `search_scope` and `search_kind` params to both the global server config, as well as to the request. Is that desirable?
2. I assume we want a `search_kind` enum instead of an OR'ed flag-like `search_kinds` to keep it simple?
3. Currently the logic still checks for the `*` and `#` markers because there might be existing LSP clients using that. Do we want it? Current precedence logic is markers -> request params -> global config
4. `config.rs` currently transforms `WorskpaceSymbolSearchScopeDef` into `lsp_ext::WorkspaceSymbolSearchScope`. I'm not sure we want an additional enum / struct indirection in between those. 
5. I haven't actually tested the per-request params case, only the VSCode setting -> rust-analyzer global config path. Is there an easy way to check the per-request case?

**Comment by flodiebold** [comment]
Since the markers work fine in Emacs, I would vote for keeping them.

**Comment by kjeremy** [comment]
Should we file an issue or proposal on https://github.com/microsoft/language-server-protocol?

File: docs/dev/lsp-extensions.md [hunk 1]
```diff
@@ -579,3 +579,16 @@ This request is sent from client to server to open the current project's Cargo.t
...
+**Issue:** https://github.com/rust-analyzer/rust-analyzer/pull/7698
+
+This request is sent from client to server to search for symbols constrained by a search
+scope and a symbol kind.
+
+**Method:** `workspace/symbol`
+
+**Request:** `WorkspaceSymbolParams`
+
+**Response:** `SymbolInformation[] | null`
```
Comment (matklad) [inline docs/dev/lsp-extensions.md:594]: Let's actually specify what we change from the stock request. Take a look at how snippet text edit is specced: 

https://github.com/matklad/rust-analyzer/blob/459b0332d3a4905fa9bc93c484538e74d191c80b/docs/dev/lsp-extensions.md#L38-L62

**Review by matklad — COMMENTED** [review]
Ouch, sorry @alcroito, this fell of my radar. Feel free to tag @matklad if the PR is not reviewed. 

The PR looks good to me as is (so, you picked the right answers for your questions!), we'd only want to expand the extension docs to actually specify the accepted format. 

>Is there an easy way to check the per-request case?

I don't think so, it's probably best to wait until ~~Emacs~~ some editor implements support for this.

File: docs/dev/lsp-extensions.md (hunk 1 above)
Comment (alcroito) [reply docs/dev/lsp-extensions.md]: Done. Hopefully the added info is enough. I also couldn't figure out whether `export` is necessary for the interfaces.

**Comment by alcroito** [comment]
Should I add some caveat / note to https://rust-analyzer.github.io/manual.html#workspace-symbol that that the `#` and '*' markers don't work with VSCode?

**Comment by matklad** [comment]
Updating manual would be helpful, yeah (it’s in this repo)

**Comment by alcroito** [comment]
Updated manual to mention VSCode users shouldn't use the filtering symbols, but should rely on the config settings instead.

**Comment by matklad** [comment]
bors r+

Thanks!

**Comment by lnicola** [comment]
bors retry

**Comment by lnicola** [comment]
bors retry

**Pull request merged**